keps/sig-scheduling/4832-async-preemption/README.md

```diff
@@ -0,0 +1,417 @@
+3. The preemption plugin blocks the Pod while the preemption routine is in-progress, using PreEnqueue extension point, so that the target Pod won't be retried during this time.
+
+Then, afterwards the preemption goroutine makes actual API calls to delete victime Pods and set `Pod.Status.NominatedNodeName`. 
+If the preemption goroutine fails at some point, it reverts the nomination via `AddNominatedPod` with [`clearNominatedNode`](https://github.com/kubernetes/kubernetes/blob/f5c538418189e119a8dbb60e2a2b22394548e326/pkg/scheduler/schedule_one.go#L135).
+If the preemption goroutine succeeds, the Pod is queued back to the queue, and (hopefully) scheduled on the nominated node.
+
+### Consideration to race condition
+
+Thanks to the nomination at `PostFilter`, this new asynchronous preemption shouldn't make any race condition between several scheduling cycles.
+
+Here, I'll discuss what happens in which scenario, and make sure there's no worry.
+
+Let's say pod1 is during the preemption process (node1) at the preemption goroutine, the next scheduling cycle is scheduling pod2. 
+
+#### The pod2's scheduling is successful (pod2 is equal or lower priority than pod1)
+
+As I described above, pod1's `PostFilter` nominates pod1 for node1.
+
+At the scheduling cycle, the scheduler takes such nominated pods that are equal or higher priority than pod1 into consideration;
+meaning, pod2 won't rob pod1 of the place on node1.
+
+#### The pod2's scheduling is successful (pod2 is higher priority than pod1)
+
+Even though pod1 is nominated for the node, the scheduler allows pod2 to take node1, where the pod1's preemption made the space. 
+
+Then, when pod1 comes back to the scheduling cycle, it may not be able to land on node1 because pod2 is scheduled there now.
+It happens with both the current and this KEP's scheduler, so no issue here.
+
+#### The pod2's scheduling is failed and starts the preemption (pod2 is equal or lower priority than pod1)
+
+The preemption also takes nominated Pods into consideration when calculating the preemption target.
+
+Therefore, if, coincidently, two preemptions for pod1 and pod2 select the same Node after all, 
+then the preemption for pod2 should decide to make the space for pod1 and pod2.
+
+So, we don't have to worry about two preemption targeting the same Node make any issue.
+
+#### The pod2's scheduling is failed and starts the preemption (pod2 is higher priority than pod1)
+
+The pod2's preemption ignores pod1's nomination for node1.
+
+If, coincidently, two preemptions for pod1 and pod2 select the same Node after all, 
+then the preemption for pod2 may just select the same preemption targets as pod1,
+and when pod1 comes back to the scheduling cycle, it (probably) cannot be scheduled on node1 because of pod2.
+
+But, this isn't an issue because the final result is completely the same as the current scheduler;
```

Comment:
This is different, because with this design, we'll delete the Pods async, during that time window, we may start the next scheduling cycle and snapshot the cluster, in a small cluster, this may race more often.

Reply:
So, here explains that it's the same result eventually even when the pod deletion isn't done by the start of next scheduling cycle, right? Is this explanation missing some edge cases?

Reply:
Yes, eventually we will get the same result. What I want to highlight is once two preemptions would like to delete the same victims (because of the async eviction), it will lead to a waste of **One** scheduling cycle and useless evictions. I think it's serious.

Reply:
So, did you want to argue the first scheduling cycle for pod2 is a waste, compared to the current scheduler, because if we wait for pod1's preemption to be done before pod2's scheduling, pod2 can go to the node straight away?

I'm not sure if it makes a serious performance difference, or could happen seriously-frequently (unless, again, I'm not missing anything).
Here's a detailed explanation (long story warning!!! as usual 😅)

The scenario that we're talking here is:
- the cluster is super packed.
- pod1's priority < pod2's priority.
- pod1 cannot find the place and triggers the preemption.
- pod2 comes after pod1, pod2 cannot find the place without any pod deletion.

And, the current scheduler behaves:
1. pod1 triggers the preemption.
2. The scheduler just waits for (1) to be done. (Note that, at this point, the victim Pods just have deletionTimestamp, not truly deleted)
3. If victim Pods are truly deleted before the scheduling cycle start, pod2's scheduling cycle starts and pod2 can go straight to node1 where pod1 makes a space with the preemption. If not, pod2 cannot find the place and goes back to the queue.

The proposed scheduler behaves:
1. pod1 triggers the preemption.
2. pod2's scheduling cycle starts without waiting for (1) to be done. And the scheduler finds that pod2 cannot go anywhere.
3. pod2 triggers the preemption and (let's say) selects the same node as pod1.

First, as written at `the current scheduler`'s third step,
note that, if not, pod2's scheduling cycle is wasted, which is completely the same as the proposed scheduler. 
_So, even the current scheduler could waste one scheduling cycle, in the first place._ 
Rather, actually, given it usually takes some time for Pods to be truly deleted, it's more likely to waste, than not to waste.

But, okay, let's consider what would happen in case the current scheduler doesn't waste; meaning all victim Pods are truly deleted luckily before the pod2's scheduling cycle start.
In this case, let's say API calls for the preemption take 1 second to be all done.
Then, the time that the scheduler takes to schedule pod2 finally is `1 second + the victim Pods are all truly deleted + the time for one scheduling cycle`.
Given "the time for one scheduling cycle" is very small, we can roughly say it's `1 second + the victim Pods are all truly deleted`.

Then, when it comes to the proposed scheduler, it'll be `1 second + max{(the victim Pods are all truly deleted), (the backoff time for pod2)} + the time for _two_ scheduling cycles.`
Given, again "the time for one scheduling cycle" is very small, we can roughly say it's 1` second + max{(the victim Pods are all truly deleted), (the backoff time for pod2)}`.

So, as a result, a rough time comparison between two schedulers would be:
- The current scheduler: `1 second + the victim Pods are all truly deleted` 
- The new scheduler: `1 second + max{(the victim Pods are all truly deleted), (the backoff time for pod2)}`.

So, if (the victim Pods are all truly deleted) > (the backoff time for pod2), then no time difference.
Also, I believe the time difference would be small, even if there is.

To summarize, if you want to observe the current scheduler is faster for pod2 than the proposed scheduler, all the following conditions have to be met:
General condition:
- the cluster is super packed.
- pod1 cannot find the place and triggers the preemption.
- pod2 comes after pod1, pod2 cannot find the place too without any pod deletion.

The current scheduler:
- pod2 has to come to the activeQ before pod1 starts a second scheduling cycle, but after all victim Pods (deleted by pod1's preemption) are truly deleted.

The proposed scheduler:
- pod2 selects the same node as pod1 in its preemption.
- pod2's backoff time is longer than the time for the victim Pods to all truly deleted.

When you're so fortunate that all the conditions are met, you can finally see the performance degradation and the difference is `(the backoff time for pod2) - (the victim Pods are all truly deleted)`, which should be not that big difference.

Reply:
Okay, looking back at my reply just made, I found my reply too long. 
So, in very short, I guess you missed that, even with the current scheduler, the Pods aren't truly deleted after the preemption (because the preemption just puts the deletionTimestamp on victim Pods). 

- It could likely make a one wasted scheduling cycle for pod2 in the current scheduler, like the proposed scheduler.
- Even if it doesn't make luckily, the time difference made by the wasted scheduling cycle in the proposed scheduler would be likely small.

Yes, there's, again, many assumptions ("likely") though, in the first place, this scenario that we're talking itself is also not something that very frequently happens in many clusters.

Reply:
Thanks Kensei, it's a long typing. With the consideration of the maturity of Kubernetes today, I believe stability is more important than new features, so I'm more concerned about what side effects would  be introduced into the kubernetes. What I mentioned above is one side, as the same victim would be chosen for two different preemptions, another one is the nominator info would be stale in the cache, for example the eviction API calls are failed, both of them may lead to chaos (I hope I worried too much)

However, I think there's no reason to block this goes into the alpha stage, and Aldo already LGTMed, but when graduating to Beta, I hope we can take these risks into consideration, so LGTM as well.
/lgtm

Reply:
> the same victim would be chosen for two different preemptions

I do not consider it as a risk both from a functional perspective and a performance perspective
as the KEP and my above too-long comment explain it's safe even if two goroutines select the same Pod(s) as a preemption victim.

> another one is the nominator info would be stale in the cache, for example the eviction API calls are failed

Yes, it is a risk in case kube-apiserver (or network etc) is unstable and the API call fails. And, we already mentioned it in the risk section.

Reply:
I wouldn't argue more on this since we're all aware of the case but if you're evicting a bunch of Pods however this wouldn't make the preemptor Pod schedulable, or even worse make the cluster less balance, I would take it a risk. We do have similar situations right now (we do not guarantee the preemptor pod will be schedulable again), but we're exacerbating this.